### PR TITLE
[flang][runtime] Added self-printing for InternalUnit.

### PR DESCRIPTION
**Pull request by vzakhari**
An InternalUnit might be constructed to allocate its own "output"
buffer of a predefined size. The buffer is then used for collecting
all the output, and it printed by std::printf at the end of the statement.

This is a suggested way for supporting 'PRINT *, ...' in the device code.
It might be not ideal, because the output is not formatted the same way
as the UNIT=5 output is formatted by default.

### CODE REVIEW

File: flang/runtime/internal-unit.h [hunk 1]
```diff
@@ -48,6 +55,10 @@ template <Direction DIR> class InternalDescriptorUnit : public ConnectionState {
   void BlankFillOutputRecord();
 
   StaticDescriptor<maxRank, true /*addendum*/> staticDescriptor_;
+  RT_OFFLOAD_VAR_GROUP_BEGIN
+  static constexpr std::size_t ownBufferSizeInBytes{1024};
+  RT_OFFLOAD_VAR_GROUP_END
```
Comment (klausler) [inline flang/runtime/internal-unit.h:60]: Should this be one line lower?

File: flang/runtime/internal-unit.h [hunk 2]
```diff
@@ -48,6 +55,10 @@ template <Direction DIR> class InternalDescriptorUnit : public ConnectionState {
...
+  bool usesOwnBuffer{false};
```
Comment (klausler) [inline flang/runtime/internal-unit.h:61]: The naming convention in the runtime (and front-end) has an underscore at the end of the name of a private `class` data member.

File: flang/runtime/io-stmt.cpp [hunk 3]
```diff
@@ -119,6 +120,9 @@ template <Direction DIR> void InternalIoStatementState<DIR>::BackspaceRecord() {
 }
 
 template <Direction DIR> int InternalIoStatementState<DIR>::EndIoStatement() {
+  if constexpr (DIR == Direction::Output) {
```
Comment (klausler) [inline flang/runtime/io-stmt.cpp:123]: If it is possible for the buffer flush to raise an error that the program can catch via `ERR=` or `IOSTAT=`, do this in `CompleteOperation` instead.

File: flang/runtime/internal-unit.cpp [hunk 4]
```diff
@@ -41,6 +52,26 @@ InternalDescriptorUnit<DIR>::InternalDescriptorUnit(
...
+      // Null terminate the buffer that contains just a single record.
+      Terminator terminator{__FILE__, __LINE__};
+      RUNTIME_CHECK(terminator,
+          furthestPositionInRecord <
+              static_cast<std::int64_t>(ownBufferSizeInBytes));
+      *reinterpret_cast<char *>(CurrentRecord() + furthestPositionInRecord) =
+          '\0';
+
+      // Print the buffer and deallocate memory.
+      // FIXME: this output does not match the regular unit 5 output.
```
Comment (klausler) [inline flang/runtime/internal-unit.cpp:67]: The default output unit is 6.

File: flang/runtime/internal-unit.cpp [hunk 5]
```diff
@@ -41,6 +52,26 @@ InternalDescriptorUnit<DIR>::InternalDescriptorUnit(
...
+      std::printf("%s\n", descriptor().OffsetElement());
+      FreeMemory(descriptor().OffsetElement());
```
Comment (klausler) [inline flang/runtime/internal-unit.cpp:69]: Why not `descriptor().Deallocate()` (and `descriptor().Allocate()`)?

**Review by klausler — COMMENTED** [review]
If you want to support child I/O (user-defined derived type I/O) on the device, this approach of adding special behavior to the internal units won't work -- child I/O is embedded in external units.  It might be clearer to support this device output path by extending the external unit class instead -- its `FileFrame` template in buffer.h can be instantiated on a `Store` that has no `Read` and only a rudimentary `Write`.

**Comment by vzakhari** [comment]
Will use alternative solution.

**Pull request closed without merging**